pkg/operator/ceph/object/user/controller.go

```diff
@@ -367,6 +368,38 @@ func generateUserConfig(user *cephv1.CephObjectStoreUser) admin.User {
 		Keys:        make([]admin.UserKeySpec, 1),
 	}
 
+	if user.Spec.MaxBuckets != 0 {
+		userConfig.MaxBuckets = &user.Spec.MaxBuckets
+	}
+
+	if user.Spec.MaxObjects != 0 || user.Spec.MaxSize != "" {
+		quotaEnable := true
+		var maxSize int64 = -1
+		var maxObjects int64 = -1
+		if user.Spec.MaxObjects >= 0 {
+			maxObjects = int64(user.Spec.MaxObjects)
+		}
+		if user.Spec.MaxSize != "" {
+			maxSizeInt, _ := resource.ParseQuantity(user.Spec.MaxSize)
+			maxSize = maxSizeInt.Value()
+		}
+		userConfig.UserQuota = admin.QuotaSpec{
+			UID:        user.Name,
+			Enabled:    &quotaEnable,
+			MaxSize:    &maxSize,
+			MaxObjects: &maxObjects,
+		}
+	}
+
+	if user.Spec.Caps != nil {
+		for capType, capPerm := range user.Spec.Caps {
+			userCap := admin.UserCapSpec{
```

Comment:
Can we validate the caps? Or if it's a well-defined list, it could also be restricted in the CRD schema.

Reply:
@travisn It is well-defined (list)[https://docs.ceph.com/en/latest/radosgw/admin/#add-remove-admin-capabilities], `--caps="[users|buckets|metadata|usage|zone]=[*|read|write|read, write]"` . I cannot quite about CRD schema for that it need to be list with multiple values, hence I choose string map. I can add validitation for the contents.

Reply:
According to [this page](https://book.kubebuilder.io/reference/markers/crd-validation.html) you should be able to restrict with a regular expression in the schema by adding `// +kubebuilder:validation:Pattern=<string>` in the types.go. Looks like we have several examples such as [this](https://github.com/rook/rook/blob/master/pkg/apis/ceph.rook.io/v1/types.go#L868).